File: scripts/scenarios/scenarios/15_mutes_relationships_starterpacks.py
"""Scenario 15: "Social Boundaries" — Mutes, Relationships & Starter Packs

Quiet Observer mutes Trollface, checks the mutes list, then unmutes.
Luna and Marcus check their relationship status. Rosa creates a starter
pack of her foodie friends.

Services: PDS, AppView (optional — some endpoints skip if no AppView)
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, ScenarioResult, timed_call



def _now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def run() -> ScenarioResult:
    result = ScenarioResult("Mutes, Relationships & Starter Packs")
    result.start()

    client = XrpcClient(PDS1)

    timed_call(result, "Server health check",
               lambda: client.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # ── Create accounts ──────────────────────────────────────────────
    char_names = ["luna", "marcus", "rosa", "troll", "quiet", "admin"]
    for name in char_names:
        char = get_character(name)
        session = timed_call(
            result, f"Create account: {char.name}",
            lambda c=char: client.accounts.create_account(c.handle, c.email, c.password),
            detail_fn=lambda s, n=name: f"did={s['did']}",
        )
        if session:
            char.did = session["did"]
            char.access_jwt = session["accessJwt"]

    active = [n for n in char_names if get_character(n).did]
    if len(active) < 4:
        result.step_failed("Account creation", "Not enough accounts")
        result.finish()
        return result

    # ── Set up profiles ──────────────────────────────────────────────
    for name in active:
        char = get_character(name)
        timed_call(result, f"Set profile: {char.name}",
                   lambda c=char: client.records.create_record(
                       c.did, "app.bsky.actor.profile",
                       {"$type": "app.bsky.actor.profile", "displayName": c.name, "description": c.persona},
                       c.access_jwt),
                   skip_on_status={404})

    luna = get_character("luna")
    marcus = get_character("marcus")
    rosa = get_character("rosa")
    troll = get_character("troll")
    quiet = get_character("quiet")

    # ── Establish follows ────────────────────────────────────────────
    for pair in [("luna", "marcus"), ("marcus", "luna")]:
        f, t = pair
        fchar = get_character(f)
        tchar = get_character(t)
        if fchar.did and tchar.did:
            timed_call(
                result, f"{fchar.name} follows {tchar.name}",
                lambda ff=fchar, tt=tchar: client.records.create_record(
                    ff.did, "app.bsky.graph.follow",
                    {"$type": "app.bsky.graph.follow", "subject": tt.did, "createdAt": _now()},
                    ff.access_jwt,
                ),
                detail_fn=lambda r: f"uri={r['uri']}",
            )

    time.sleep(1)

    # ── Quiet mutes Trollface ────────────────────────────────────────
    if quiet.did and quiet.access_jwt and troll.did:
        timed_call(
            result, "Quiet mutes Trollface",
            lambda: client.graph.mute_actor(troll.did, quiet.access_jwt),
        )
    else:
        result.step_skipped("Quiet mutes Trollface", "Missing DID or token")

    # ── Quiet checks mutes list ──────────────────────────────────────
    if quiet.access_jwt:
        timed_call(
            result, "Quiet checks mutes list",
            lambda: client.graph.get_mutes(quiet.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('mutes', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Quiet checks mutes list", "Quiet not created")

    # ── Luna checks mutes (should be empty) ─────────────────────────
    if luna.access_jwt:
        timed_call(
            result, "Luna checks mutes (empty)",
            lambda: client.graph.get_mutes(luna.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('mutes', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Luna checks mutes (empty)", "Luna not created")

    # ── Quiet unmutes Trollface ──────────────────────────────────────
    if quiet.did and quiet.access_jwt and troll.did:
        timed_call(
            result, "Quiet unmutes Trollface",
            lambda: client.graph.unmute_actor(troll.did, quiet.access_jwt),
        )
    else:
        result.step_skipped("Quiet unmutes Trollface", "Missing DID or token")

    # ── Quiet verifies Trollface removed from mutes ─────────────────
    if quiet.access_jwt:
        timed_call(
            result, "Quiet verifies unmute",
            lambda: client.graph.get_mutes(quiet.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('mutes', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Quiet verifies unmute", "Quiet not created")

    # ── Check relationships ──────────────────────────────────────────
    if luna.did and marcus.did:
        timed_call(
            result, "Luna→Marcus relationship",
            lambda: client.graph.get_relationships(luna.did, [marcus.did], token=luna.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('relationships', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Luna→Marcus relationship", "Missing DIDs")

    if luna.did and troll.did:
        timed_call(
            result, "Luna→Troll relationship",
            lambda: client.graph.get_relationships(luna.did, [troll.did], token=luna.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('relationships', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Luna→Troll relationship", "Missing DIDs")

    # ── Rosa creates a starter pack ──────────────────────────────────
    rosa_sp_uri = None
    if rosa.did and rosa.access_jwt and luna.did and marcus.did:
        sp = timed_call(
            result, "Rosa creates starter pack",
            lambda: client.records.create_record(
                rosa.did, "app.bsky.graph.starterpack",
                {
                    "$type": "app.bsky.graph.starterpack",
                    "name": "Foodie Friends",
                    "description": "My favorite food-loving friends!",
                    "createdAt": _now(),
                },
                rosa.access_jwt,
            ),
            detail_fn=lambda r: f"uri={r['uri']}",
        )
        if sp:
            rosa_sp_uri = sp["uri"]
    else:
        result.step_skipped("Rosa creates starter pack", "Missing DIDs")

    # ── Get Rosa's starter packs ─────────────────────────────────────
    if rosa.did:
        timed_call(
            result, "Rosa's starter packs",
            lambda: client.graph.get_actor_starter_packs(rosa.did, token=rosa.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('starterPacks', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Rosa's starter packs", "Rosa not created")

    # ── Get specific starter pack by URI ─────────────────────────────
    if rosa_sp_uri:
        timed_call(
            result, "Get starter pack by URI",
            lambda: client.graph.get_starter_pack(rosa_sp_uri, token=rosa.access_jwt),
            detail_fn=lambda r: f"name={r.get('starterPack', {}).get('name', '')}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Get starter pack by URI", "No starter pack URI")

    # ── Get multiple starter packs by URIs ───────────────────────────
    if rosa_sp_uri:
        timed_call(
            result, "Get starter packs by URIs",
            lambda: client.graph.get_starter_packs([rosa_sp_uri], token=rosa.access_jwt),
            detail_fn=lambda r: f"count={len(r.get('starterPacks', []))}",
            skip_on_status={404},
        )
    else:
        result.step_skipped("Get starter packs by URIs", "No starter pack URI")

    # ── Record artifacts ─────────────────────────────────────────────
    result.record_artifact("accounts", {
        n: {"did": get_character(n).did}
        for n in char_names if get_character(n).did
    })
    result.record_artifact("relationships_checked", ["luna→marcus", "luna→troll"])
    result.record_artifact("starter_pack_uri", rosa_sp_uri)

    result.finish()
    return result


if __name__ == "__main__":
    result = run()
    result.print_summary()
    sys.exit(result.exit_code)

File: scripts/scenarios/scenarios/15_mutes_relationships_starterpacks.ts
/**
 * @module scenarios/15_mutes_relationships_starterpacks
 *
 * Scenario: Mutes, Relationships, and Starter Packs Management
 *
 * Behavior:
 * - Creates multiple user accounts and profiles.
 * - Establishes follow relationships between users.
 * - Tests mute/unmute functionality for actors.
 * - Verifies relationship lookups.
 * - Tests creating and retrieving lists and starter packs.
 *
 * Expectations:
 * - Relationships and mutes are tracked correctly.
 * - Starter packs can be created and retrieved by URI.
 * - User interactions adhere to the expected graph API behavior.
 */

import { ScenarioResult, timedCall } from "../../lib/deno/runner.ts";
export { ScenarioResult, StepResult, StepStatus } from "../../lib/deno/runner.ts";
export type { ScenarioReport } from "../../lib/deno/runner.ts";
import { assert } from "../../lib/deno/assertions.ts";
import { XrpcClient, XrpcError } from "../../lib/deno/client.ts";
import { getActor, PDS1 } from "../../lib/deno/config.ts";

function now() {
  return new Date().toISOString();
}

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */
export async function run(): Promise<ScenarioResult> {
  const result = new ScenarioResult("Mutes, Relationships & Starter Packs");
  result.start();

  const client = new XrpcClient(PDS1);

  await timedCall(result, "Server health check", async () => {
    await client.waitForHealthy(30);
  });

  if (result.failed > 0) {
    result.finish();
    return result;
  }

  const charNames = ["luna", "marcus", "rosa", "troll", "quiet", "admin"];
  for (const name of charNames) {
    const char = getActor(name);
    const session = await timedCall(
      result,
      `Create account: ${char.name}`,
      async () => {
        return await client.accounts.createAccount(char.handle, char.email, char.password);
      },
      (s) => `did=${s.did}`,
    );
    if (session) {
      char.did = session.did;
      char.accessJwt = session.accessJwt;
    }
  }

  const active = charNames.filter((n) => getActor(n).did);
  for (const name of active) {
    const char = getActor(name);
    try {
      await client.records.createRecord(
        char.did,
        "app.bsky.actor.profile",
        { $type: "app.bsky.actor.profile", displayName: char.name, description: char.persona },
        char.accessJwt,
      );
    } catch (e) {
      if (!(e instanceof XrpcError && e.status === 404)) throw e;
    }
  }

  const luna = getActor("luna");
  const marcus = getActor("marcus");
  const rosa = getActor("rosa");
  const troll = getActor("troll");
  const quiet = getActor("quiet");

  // Establish follows
  for (const [f, t] of [["luna", "marcus"], ["marcus", "luna"]] as const) {
    const ff = getActor(f);
    const tt = getActor(t);
    if (ff.did && tt.did) {
      await timedCall(
        result,
        `${ff.name} follows ${tt.name}`,
        async () => {
          return await client.records.createRecord(
            ff.did,
            "app.bsky.graph.follow",
            { $type: "app.bsky.graph.follow", subject: tt.did, createdAt: now() },
            ff.accessJwt,
          );
        },
        (r) => `uri=${r.uri}`,
      );
    }
  }

  await new Promise((r) => setTimeout(r, 1000));

  if (quiet.did && troll.did) {
    await timedCall(
      result,
      "Quiet mutes Trollface",
      async () => {
        return await client.graph.muteActor(troll.did, quiet.accessJwt);
      },
    );

    await timedCall(
      result,
      "Quiet checks mutes list",
      async () => {
        return await client.graph.getMutes(quiet.accessJwt);
      },
      (r) => `count=${r.mutes?.length || 0}`,
    );

    await timedCall(
      result,
      "Quiet unmutes Trollface",
      async () => {
        return await client.graph.unmuteActor(troll.did, quiet.accessJwt);
      },
    );

    await timedCall(
      result,
      "Quiet verifies unmute",
      async () => {
        return await client.graph.getMutes(quiet.accessJwt);
      },
      (r) => `count=${r.mutes?.length || 0}`,
    );
  }

  if (luna.did && marcus.did) {
    await timedCall(
      result,
      "Luna→Marcus relationship",
      async () => {
        return await client.graph.getRelationships(luna.did, [marcus.did], luna.accessJwt);
      },
      (r) => `count=${r.relationships?.length || 0}`,
    );
  }

  let rosaSpUri: string | null = null;
  if (rosa.did && luna.did && marcus.did) {
    const listRec = await timedCall(
      result,
      "Rosa creates list for starter pack",
      async () => {
        return await client.records.createRecord(
          rosa.did,
          "app.bsky.graph.list",
          {
            $type: "app.bsky.graph.list",
            name: "Foodie Friends",
            purpose: "app.bsky.graph.defs#curatelist",
            createdAt: now(),
          },
          rosa.accessJwt,
        );
      },
      (r) => `uri=${r.uri}`,
    );

    if (listRec) {
      const sp = await timedCall(
        result,
        "Rosa creates starter pack",
        async () => {
          return await client.records.createRecord(
            rosa.did,
            "app.bsky.graph.starterpack",
            {
              $type: "app.bsky.graph.starterpack",
              name: "Foodie Friends",
              list: listRec.uri,
              createdAt: now(),
            },
            rosa.accessJwt,
          );
        },
        (r) => `uri=${r.uri}`,
      );
      rosaSpUri = sp?.uri || null;
    }
  }

  if (rosa.did) {
    await timedCall(
      result,
      "Wait for starter pack in AppView",
      async () => {
        const deadline = Date.now() + 30_000;
        while (Date.now() < deadline) {
          const r = await client.graph.getActorStarterPacks(rosa.did, { token: rosa.accessJwt });
          if (r.starterPacks && r.starterPacks.length > 0) return r;
          await new Promise((r) => setTimeout(r, 1000));
        }
        throw new Error("Timed out waiting for starter pack to appear in AppView");
      },
      (r) => `count=${r.starterPacks?.length || 0}`,
    );
  }

  if (rosaSpUri) {
    await timedCall(
      result,
      "Get starter pack by URI",
      async () => {
        return await client.graph.getStarterPack(rosaSpUri!, rosa.accessJwt);
      },
      (r) => `name=${r.starterPack?.name || ""}`,
    );

    await timedCall(
      result,
      "Get starter packs by URIs",
      async () => {
        return await client.graph.getStarterPacks([rosaSpUri!], rosa.accessJwt);
      },
      (r) => `count=${r.starterPacks?.length || 0}`,
    );
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run().then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
